nni/retiarii/hub/pytorch/autoformer.py

```diff
@@ -146,91 +139,153 @@ class TransformerEncoderLayer(nn.Module):
+        x = F.dropout(x, p=self.dropout, training=self.training)
         x = self.fc2(x)
-        x = nn.functional.dropout(x, p=self.dropout, training=self.training)
+        x = F.dropout(x, p=self.dropout, training=self.training)
         x = self.drop_path(x)
         x = residual + x
         x = self.maybe_layer_norm(self.ffn_layer_norm, x, after=True)
+
         return x
 
-    def maybe_layer_norm(self, layer_norm, x, before=False, after=False):
-        assert before ^ after
-        if after ^ self.normalize_before:
-            return layer_norm(x)
-        else:
-            return x
+
+@basic_unit
+class ClsToken(nn.Module):
+    """ Concat class token with dim=embed_dim before patch embedding. 
+    """
+    def __init__(self, embed_dim: int):
+        super().__init__()
+        self.cls_token = nn.Parameter(torch.zeros(1, 1, embed_dim))
+        trunc_normal_(self.cls_token, std=.02)
+    
+    def forward(self, x):
+        return torch.cat((self.cls_token.expand(x.shape[0], -1, -1), x), dim=1)
+
+
+class MixedClsToken(MixedOperation, ClsToken):
+    """ Mixed class token concat operation.
+
+    Supported arguments are:
+
+    - ``embed_dim``
+    
+    Prefix of cls_token will be sliced.
+    """
+    bound_type = ClsToken
+    argument_list = ['embed_dim']
+
+    def super_init_argument(self, name: str, value_choice: ValueChoiceX):
+        return max(traverse_all_options(value_choice))
+
+    def forward_with_args(self, embed_dim, 
+                        inputs: torch.Tensor) -> torch.Tensor:
+        cls_token = self.cls_token[..., :embed_dim]
+
+        return torch.cat((cls_token.expand(inputs.shape[0], -1, -1), inputs), dim=1)
+
+
+@basic_unit
+class AbsPosEmbed(nn.Module):
+    """ Add absolute position embedding on patch embedding.
+    """
+    def __init__(self, length: int, embed_dim: int):
+        super().__init__()
+        self.pos_embed = nn.Parameter(torch.zeros(1, length, embed_dim))
+        trunc_normal_(self.pos_embed, std=.02)
+
+    def forward(self, x):
+        return x + self.pos_embed
+
+
+class MixedAbsPosEmbed(MixedOperation, AbsPosEmbed):
+    """ Mixed absolute position embedding add operation.
+
+    Supported arguments are:
+
+    - ``embed_dim``
+    
+    Prefix of pos_embed will be sliced.
+    """
+    bound_type = AbsPosEmbed
+    argument_list = ['embed_dim']
+
+    def super_init_argument(self, name: str, value_choice: ValueChoiceX):
+        return max(traverse_all_options(value_choice))
+
+    def forward_with_args(self,  embed_dim, 
+                        inputs: torch.Tensor) -> torch.Tensor:
+        pos_embed = self.pos_embed[..., :embed_dim]
+
+        return inputs + pos_embed
 
 
 @model_wrapper
 class AutoformerSpace(nn.Module):
     """
     The search space that is proposed in `Autoformer <https://arxiv.org/abs/2107.00651>`__.
     There are four searchable variables: depth, embedding dimension, heads number and MLP ratio.
-
```

Comment:
Actually, you must have this blank line. Otherwise doc is not properly rendered.